internal/core/software_renderer.rs

```diff
@@ -314,12 +314,73 @@ impl Renderer for SoftwareRenderer {
     ) -> usize {
         0
     }
+
     fn text_input_cursor_rect_for_byte_offset(
         &self,
-        _text_input: Pin<&crate::items::TextInput>,
-        _byte_offset: usize,
+        text_input: Pin<&crate::items::TextInput>,
+        byte_offset: usize,
     ) -> LogicalRect {
-        Default::default()
+        let window_adapter = match self.window.upgrade() {
+            Some(window) => window,
+            None => return Default::default(),
+        };
+
+        let scale_factor = ScaleFactor::new(window_adapter.window().scale_factor()).cast();
+        let string = text_input.text();
+
+        let font_request = text_input.font_request(&window_adapter);
+        let font = fonts::match_font(&font_request, scale_factor);
+
+        let width = (text_input.width().cast() * scale_factor).cast();
+        let height = (text_input.height().cast() * scale_factor).cast();
+
+        let (cursor_position, cursor_height) = match font {
+            fonts::Font::PixelFont(pf) => {
+                let layout = fonts::text_layout_for_font(&pf, &font_request, scale_factor);
+
+                let paragraph = TextParagraphLayout {
+                    string: &string,
+                    layout,
+                    max_width: width,
+                    max_height: height,
+                    horizontal_alignment: text_input.horizontal_alignment(),
+                    vertical_alignment: text_input.vertical_alignment(),
+                    wrap: text_input.wrap(),
+                    overflow: TextOverflow::Clip,
+                    single_line: false,
+                };
+
+                (paragraph.cursor_pos_for_byte_offset(byte_offset), pf.height())
+            }
+            #[cfg(feature = "software-renderer-systemfonts")]
+            fonts::Font::VectorFont(vf) => {
+                let layout = fonts::text_layout_for_font(&vf, &font_request, scale_factor);
+
+                let paragraph = TextParagraphLayout {
+                    string: &string,
+                    layout,
+                    max_width: width,
+                    max_height: height,
+                    horizontal_alignment: text_input.horizontal_alignment(),
+                    vertical_alignment: text_input.vertical_alignment(),
+                    wrap: text_input.wrap(),
+                    overflow: TextOverflow::Clip,
+                    single_line: false,
+                };
+
+                (paragraph.cursor_pos_for_byte_offset(byte_offset), vf.height())
+            }
+        };
+
+        return PhysicalRect::new(
+            PhysicalPoint::from_lengths(cursor_position.0, cursor_position.1),
+            PhysicalSize::from_lengths(
+                (text_input.text_cursor_width().cast() * scale_factor).cast(),
+                cursor_height,
+            ),
+        )
+        .cast()
+        .cast_unit();
```

Comment:
This can't be correct. This constructs `PhysicalRect` but the function is supposed to return a `LogicaRect`. `cast_unit()` just casts the unit but preserves the value. So either this should construct a `LogicalRect` from the beginning, or another cast is needed.